x-pack/plugins/monitoring/server/alerts/memory_usage_alert.ts

```diff
@@ -0,0 +1,343 @@
+  ): Promise<AlertData[]> {
+    let esIndexPattern = appendMetricbeatIndex(this.config, INDEX_PATTERN_ELASTICSEARCH);
+    if (availableCcs) {
+      esIndexPattern = getCcsIndexPattern(esIndexPattern, availableCcs);
+    }
+    const { duration, threshold } = params;
+    const stats = await fetchMemoryUsageNodeStats(
+      callCluster,
+      clusters,
+      esIndexPattern,
+      duration as string,
+      this.config.ui.max_bucket_size
+    );
+
+    return stats.map((stat) => {
+      const { clusterUuid, nodeId, memoryUsage, ccs } = stat;
+      return {
+        instanceKey: `${clusterUuid}:${nodeId}`,
+        shouldFire: memoryUsage > threshold,
+        severity: AlertSeverity.Danger,
+        meta: stat,
+        clusterUuid,
+        ccs,
+      };
+    });
+  }
+
+  protected filterAlertInstance(alertInstance: RawAlertInstance, filters: CommonAlertFilter[]) {
+    const alertInstanceStates = alertInstance.state?.alertStates as AlertMemoryUsageState[];
+    const nodeFilter = filters?.find((filter) => filter.nodeUuid);
+
+    if (!filters || !filters.length || !alertInstanceStates?.length || !nodeFilter?.nodeUuid) {
+      return true;
+    }
+
+    const nodeAlerts = alertInstanceStates.filter(({ nodeId }) => nodeId === nodeFilter.nodeUuid);
+    return Boolean(nodeAlerts.length);
+  }
+
+  protected getDefaultAlertState(cluster: AlertCluster, item: AlertData): AlertState {
+    const currentState = super.getDefaultAlertState(cluster, item);
+    currentState.ui.severity = AlertSeverity.Warning;
+    return currentState;
+  }
+
+  protected getUiMessage(alertState: AlertState, item: AlertData): AlertMessage {
+    const stat = item.meta as AlertMemoryUsageState;
+    if (!alertState.ui.isFiring) {
+      return {
+        text: i18n.translate('xpack.monitoring.alerts.memoryUsage.ui.resolvedMessage', {
+          defaultMessage: `The JVM memory usage on node {nodeName} is now under the threshold, currently reporting at {memoryUsage}% as of #resolved`,
+          values: {
+            nodeName: stat.nodeName,
+            memoryUsage: stat.memoryUsage.toFixed(2),
+          },
+        }),
+        tokens: [
+          {
+            startToken: '#resolved',
+            type: AlertMessageTokenType.Time,
+            isAbsolute: true,
+            isRelative: false,
+            timestamp: alertState.ui.resolvedMS,
+          } as AlertMessageTimeToken,
+        ],
+      };
+    }
+    return {
+      text: i18n.translate('xpack.monitoring.alerts.memoryUsage.ui.firingMessage', {
+        defaultMessage: `Node #start_link{nodeName}#end_link is reporting JVM memory usage of {memoryUsage}% at #absolute`,
+        values: {
+          nodeName: stat.nodeName,
+          memoryUsage: stat.memoryUsage,
+        },
+      }),
+      nextSteps: [
+        createLink(
+          i18n.translate('xpack.monitoring.alerts.memoryUsage.ui.nextSteps.tuneThreadPools', {
+            defaultMessage: '#start_linkTune thread pools#end_link',
```

Comment:
I don't see this listed in Ravi's doc. Was it a last minute addition?

Reply:
It wasn't on the list, but based on my research it is also a valuable resource to mitigate high memory usage. cc: @ravikesarwani wdyt?